### PR TITLE
 fix `hypot`, return `Wirtinger` appropriately

### PR DESCRIPTION
**Pull request by simeonschaub**
Turns out the derivative for `hypot` was wrong, even in the real case. Now supports an arbitrary number of real or complex arguments. Also in the process discovered that the `rrule` for `*` was wrong for complex arguments. ~~Still needs tests.~~

### CODE REVIEW

**Comment by simeonschaub** [comment]
Bump. Can we get this merged?

File: src/rules/base.jl [hunk 1]
```diff
@@ -37,7 +37,7 @@
 @scalar_rule(conj(x), Wirtinger(Zero(), One()))
 @scalar_rule(adjoint(x), Wirtinger(Zero(), One()))
 @scalar_rule(transpose(x), One())
-@scalar_rule(abs(x), sign(x))
+@scalar_rule(abs(x), x isa Real ? sign(x) : Wirtinger(x' / 2Ω, x / 2Ω))
```
Comment (ararslan) [inline src/rules/base.jl]:
```suggestion
@scalar_rule(abs(x::Real), sign(x))
@scalar_rule(abs(x::Complex), Wirtinger(x' / 2Ω, x / 2Ω))
```

File: src/rules/base.jl [hunk 2]
```diff
@@ -62,7 +62,37 @@
 @scalar_rule(cotd(x), -(π / oftype(x, 180)) * (1 + Ω^2))
 @scalar_rule(sech(x), -tanh(x) * Ω)
 @scalar_rule(csch(x), -coth(x) * Ω)
-@scalar_rule(hypot(x, y), (y / Ω, x / Ω))
+
+function frule(::typeof(hypot), x::Real...)
+    Ω = hypot(x...)
+    return Ω, Rule((Δ...) -> sum(Δ .* x) * inv(Ω))
```
Comment (ararslan) [inline src/rules/base.jl]:
```suggestion
    return Ω, Rule((Δ...) -> sum(Δ .* x) / Ω)
```

File: src/rules/base.jl [hunk 3]
```diff
@@ -62,7 +62,37 @@
...
+    Ω = hypot(x...)
+    return Ω, map(x) do x_i
+        Rule(ΔΩ -> ΔΩ * (x_i / Ω))
+    end
+end
+
+function frule(::typeof(hypot), x...)
+    Ω = hypot(x...)
+    return Ω, WirtingerRule(
+        Rule((Δ...) -> sum(Δ .* conj.(x) * inv(2Ω))),
```
Comment (ararslan) [inline src/rules/base.jl]: This should be moved flush with the opening parenthesis with the second line aligned below it.
```julia
    return Ω, WirtingerRule(Rule((Δ...) -> sum(Δ .* conj.(x) * inv(2Ω))),
                            Rule((Δ...) -> sum(Δ .* x) * inv(2Ω)))
```
Again in this case, why multiply by `inv` instead of just dividing?

File: test/rules/base.jl [hunk 4]
```diff
@@ -1,10 +1,27 @@
+_isapprox(x, y; kwargs...) = isapprox(x, y; kwargs...)
```
Comment (ararslan) [inline test/rules/base.jl]: Same comments as in the other PR.

File: test/rules/base.jl [hunk 5]
```diff
@@ -65,36 +82,46 @@ end
         # TODO: atan2 sincos
     end
     @testset "Misc. Tests" begin
-        @testset "*(x, y)" begin
-            x, y = rand(3, 2), rand(2, 5)
+        @testset "*(x, y)" for T in (Float32, Float64, Complex{Float64})
```
Comment (ararslan) [inline test/rules/base.jl]:
```suggestion
        @testset "*(x::$T, y::$T)" for T in (Float32, Float64, Complex{Float64})
```
otherwise the name of the testset doesn't actually tell you what it's testing.

File: src/rules/base.jl (hunk 2 above)
Comment (simeonschaub) [reply src/rules/base.jl]: The problem is that `Zero() / x` and `One() / x` doesn't get overdubbed, so this throws errors if all `Δ`s are `Zero()`.

File: src/rules/base.jl (hunk 3 above)
Comment (simeonschaub) [reply src/rules/base.jl]: See my comment above. It might make sense to overload `/` for `Zero()` and `One()` though.

File: test/rules/base.jl (hunk 5 above)
Comment (simeonschaub) [reply test/rules/base.jl]: Thanks for the suggestion! Will change that.

File: src/rules/base.jl (hunk 2 above)
Comment (ararslan) [reply src/rules/base.jl]: Is that because only multiplications and additions are overdubbed?

Comment (simeonschaub) [reply src/rules/base.jl]: Yes, exactly

File: src/rules/base.jl (hunk 3 above)
Comment (ararslan) [reply src/rules/base.jl]: Yeah I think that would probably make sense.

File: src/rules/base.jl (hunk 1 above)
Comment (simeonschaub) [reply src/rules/base.jl]: Now that we have that, this is definitely nicer!

File: src/rules/base.jl (hunk 3 above)
Comment (simeonschaub) [reply src/rules/base.jl]: Would it be ok if I left it for now? I'm not too experienced with Cassette and there would be some edge cases to think about, so I think this deserves its own PR.

File: test/rules/base.jl (hunk 4 above)
Comment (simeonschaub) [reply test/rules/base.jl]: I'll delete it once #64 is merged

File: src/rules/base.jl (hunk 3 above)
Comment (simeonschaub) [reply src/rules/base.jl]: I had a go at it. Could maybe be done more elegantly, but works for now.
